src/services/transaction_watcher_service.ts

```diff
@@ -0,0 +1,117 @@
+import { intervalUtils } from '@0x/utils';
+import { SupportedProvider, Web3Wrapper } from '@0x/web3-wrapper';
+import { Connection, Not, Repository } from 'typeorm';
+
+import { TX_WATCHER_POLLING_INTERVAL_MS } from '../constants';
+import { TransactionEntity } from '../entities';
+import { logger } from '../logger';
+import { TransactionStates } from '../types';
+
+export class TransactionWatcherService {
+    private readonly _transactionRepository: Repository<TransactionEntity>;
+    private readonly _provider: SupportedProvider;
+    private readonly _web3Wrapper: Web3Wrapper;
+    private readonly _transactionWatcherTimer: NodeJS.Timer;
+
+    constructor(dbConnection: Connection, provider: SupportedProvider) {
+        this._transactionRepository = dbConnection.getRepository(TransactionEntity);
+        this._provider = provider;
+        this._web3Wrapper = new Web3Wrapper(this._provider);
+        this._transactionWatcherTimer = intervalUtils.setAsyncExcludingInterval(
+            async () => {
+                logger.trace('syncing transaction status');
+                await this.syncTransactionStatusAsync();
+            },
+            TX_WATCHER_POLLING_INTERVAL_MS,
+            (err: Error) => {
+                logger.error({ message: `order watcher failed to sync transaction status`, err: err.stack });
+            },
+        );
+    }
+    public stop(): void {
+        intervalUtils.clearAsyncExcludingInterval(this._transactionWatcherTimer);
+    }
+    public async syncTransactionStatusAsync(): Promise<void> {
+        const transactionsToCheck = await this._transactionRepository.find({
+            where: [
+                { status: TransactionStates.Unsubmitted },
+                { status: TransactionStates.Submitted },
+                { status: TransactionStates.Mempool },
+                { status: TransactionStates.Stuck },
+            ],
+        });
+        logger.trace(`found ${transactionsToCheck.length} transactions to check status`);
+        for (const tx of transactionsToCheck) {
+            await this._findTransactionStatusAndUpdateAsync(tx);
+        }
+    }
+    private async _findTransactionStatusAndUpdateAsync(txEntity: TransactionEntity): Promise<TransactionEntity> {
+        // TODO(oskar) - LessThanOrEqual and LessThan do not work on dates in
+        // TypeORM queries, ref: https://github.com/typeorm/typeorm/issues/3959
+        const now = new Date();
+        const isExpired = txEntity.expectedAt <= now;
```

Comment:
I would suggest `txEntity.expectedAt <= latestBlock.timestamp`

Reply:
👍 changed. We currently store `expectedAt` as `Date` so I went with converting `latestBlock.timestamp` to a date object.